// tubular-react-common/dist/state/actions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.actions = exports.INIT_GRID_FROM_STORAGE = exports.UPDATE_SEARCH_TEXT = exports.UPDATE_ITEMS_PER_PAGE = exports.SET_COLUMNS = exports.GOTO_PAGE = exports.REQUEST_ERROR = exports.REQUEST_SUCCESS = exports.START_REQUEST = void 0;
var actionHelpers_1 = require("./actionHelpers");
exports.START_REQUEST = 'start request';
exports.REQUEST_SUCCESS = 'request success';
exports.REQUEST_ERROR = 'request error';
exports.GOTO_PAGE = 'goto page';
exports.SET_COLUMNS = 'set columns';
exports.UPDATE_ITEMS_PER_PAGE = 'update items per page';
exports.UPDATE_SEARCH_TEXT = 'update search text';
exports.INIT_GRID_FROM_STORAGE = 'init grid from storage';
exports.actions = {
    startRequest: function () { return actionHelpers_1.createAction(exports.START_REQUEST); },
    goToPage: function (payload) { return actionHelpers_1.createAction(exports.GOTO_PAGE, payload); },
    requestDone: function (payload) { return actionHelpers_1.createAction(exports.REQUEST_SUCCESS, payload); },
    requestError: function (payload) { return actionHelpers_1.createAction(exports.REQUEST_ERROR, payload); },
    setColumns: function (payload) { return actionHelpers_1.createAction(exports.SET_COLUMNS, payload); },
    updateItemsPerPage: function (payload) { return actionHelpers_1.createAction(exports.UPDATE_ITEMS_PER_PAGE, payload); },
    updateSearchText: function (payload) { return actionHelpers_1.createAction(exports.UPDATE_SEARCH_TEXT, payload); },
    initGridFromStorage: function (payload) {
        return actionHelpers_1.createAction(exports.INIT_GRID_FROM_STORAGE, payload);
    },
};
