// tubular-react-common/dist/useTubular.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTubular = void 0;
var React = require("react");
var tubular_common_1 = require("tubular-common");
var helpers_1 = require("./helpers");
var actions_1 = require("./state/actions");
var reducer_1 = require("./state/reducer");
var createTbOptions = function (tubularOptions) {
    var temp = tubularOptions || {};
    return {
        callbacks: temp.callbacks || {
            onError: function () {
                return;
            },
        },
        componentName: temp.componentName || helpers_1.tbId(),
        deps: temp.deps || null,
        pagination: temp.pagination || {
            itemsPerPage: 10,
            page: 0,
        },
        searchText: temp.searchText || '',
        storage: (temp.componentName && temp.storage) || new tubular_common_1.NullStorage(),
    };
};
var useTubular = function (initColumns, source, tubularOptions) {
    var tbOptions = createTbOptions(tubularOptions);
    var componentName = tbOptions.componentName, pagination = tbOptions.pagination, callbacks = tbOptions.callbacks, storage = tbOptions.storage, deps = tbOptions.deps, searchText = tbOptions.searchText;
    var initStorage = storage || new tubular_common_1.NullStorage();
    if (initStorage instanceof tubular_common_1.LocalStorage) {
        initStorage.setGridName(componentName);
    }
    var _a = React.useReducer(reducer_1.tbReducer, __assign(__assign({}, reducer_1.tbInitialState), { columns: initColumns, page: pagination.page || 0, itemsPerPage: pagination.itemsPerPage || 10, searchText: searchText || '' })), tbState = _a[0], dispatch = _a[1];
    var getStorage = React.useState(initStorage)[0];
    var _b = React.useState(0), refresh = _b[0], setRefresh = _b[1];
    var forceRefresh = function () { return setRefresh(function (refresh) { return refresh + 1; }); };
    var getAllRecords = React.useCallback(function () {
        return source instanceof Array ? helpers_1.getLocalDataSource(source) : helpers_1.getRemoteDataSource(source);
    }, [source]);
    var api = {
        exportTo: function (allRows, exportFunc) { return __awaiter(void 0, void 0, void 0, function () {
            var payload, _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (tbState.filteredRecordCount === 0) {
                            return [2 /*return*/];
                        }
                        if (!allRows) return [3 /*break*/, 2];
                        return [4 /*yield*/, getAllRecords()(new tubular_common_1.GridRequest(tbState.columns, -1, 0, tbState.searchText))];
                    case 1:
                        _a = (_b.sent()).payload;
                        return [3 /*break*/, 3];
                    case 2:
                        _a = tbState.data;
                        _b.label = 3;
                    case 3:
                        payload = _a;
                        exportFunc(payload, tbState.columns);
                        return [2 /*return*/];
                }
            });
        }); },
        goToPage: function (page) {
            dispatch(actions_1.actions.goToPage(page));
        },
        processRequest: function () { return __awaiter(void 0, void 0, void 0, function () {
            var request, response, maxPage, currentPage, err_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        dispatch(actions_1.actions.startRequest());
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        request = new tubular_common_1.GridRequest(tbState.columns, tbState.itemsPerPage, tbState.page, tbState.searchText);
                        return [4 /*yield*/, getAllRecords()(request)];
                    case 2:
                        response = _a.sent();
                        maxPage = Math.ceil(response.totalRecordCount / tbState.itemsPerPage);
                        currentPage = response.currentPage > maxPage ? maxPage : response.currentPage;
                        currentPage = currentPage === 0 ? 0 : currentPage - 1;
                        dispatch(actions_1.actions.requestDone({
                            page: currentPage,
                            aggregate: response.aggregationPayload,
                            data: response.payload,
                            filteredRecordCount: response.filteredRecordCount || 0,
                            totalRecordCount: response.totalRecordCount || 0,
                            columns: tbState.columns,
                            error: null,
                            searchText: tbState.searchText,
                        }));
                        return [3 /*break*/, 4];
                    case 3:
                        err_1 = _a.sent();
                        if (callbacks.onError) {
                            callbacks.onError(err_1);
                        }
                        dispatch(actions_1.actions.requestError(err_1));
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/];
                }
            });
        }); },
        reloadGrid: function (resetPage) {
            if (resetPage === void 0) { resetPage = false; }
            if (resetPage) {
                dispatch(actions_1.actions.goToPage(0));
            }
            forceRefresh();
        },
        setColumns: function (columns) { return dispatch(actions_1.actions.setColumns(columns)); },
        sortColumn: function (property, multiSort) {
            if (multiSort === void 0) { multiSort = false; }
            var columns = tubular_common_1.sortColumnArray(property, __spreadArrays(tbState.columns), multiSort);
            dispatch(actions_1.actions.setColumns(columns));
        },
        updateItemsPerPage: function (itemsPerPage) {
            if (tbState.itemsPerPage !== itemsPerPage) {
                dispatch(actions_1.actions.updateItemsPerPage(itemsPerPage));
            }
        },
        updateSearchText: function (value) {
            if (tbState.searchText !== value) {
                dispatch(actions_1.actions.updateSearchText(value));
            }
        },
    };
    var initGrid = function () {
        var initData = {
            page: tbState.page,
            searchText: tbState.searchText,
            columns: tbState.columns,
        };
        if (getStorage.getPage()) {
            initData.page = getStorage.getPage();
        }
        if (getStorage.getTextSearch()) {
            initData.searchText = getStorage.getTextSearch();
        }
        var storedColumns = getStorage.getColumns();
        if (storedColumns) {
            var columns_1 = __spreadArrays(tbState.columns);
            storedColumns.forEach(function (column) {
                var currentColumn = columns_1.find(function (col) { return col.name === column.name; });
                if (!currentColumn) {
                    return;
                }
                currentColumn.visible = column.visible;
                if (currentColumn.filterText !== null && column.filterOperator !== tubular_common_1.CompareOperators.None) {
                    return;
                }
                currentColumn.filterText = column.filterText;
                currentColumn.filterOperator = column.filterOperator;
                currentColumn.filterArgument = column.filterArgument;
            });
            initData.columns = columns_1;
        }
        dispatch(actions_1.actions.initGridFromStorage(initData));
    };
    var extraDependencies = deps ? deps : [];
    React.useEffect(function () {
        if (!tbState.initialized) {
            initGrid();
        }
    }, []);
    React.useEffect(function () {
        if (!tbState.isLoading && tbState.initialized) {
            api.processRequest();
        }
    }, __spreadArrays([
        tbState.columns,
        tbState.page,
        tbState.searchText,
        tbState.itemsPerPage,
        tbState.initialized,
        refresh,
        source
    ], extraDependencies));
    React.useEffect(function () {
        dispatch(actions_1.actions.setColumns(initColumns));
    }, [initColumns]);
    var tbInstance = {
        state: tbState,
        api: api,
    };
    return tbInstance;
};
exports.useTubular = useTubular;
