// @material-ui/pickers/dimensions-3ff816e9.js
import { useContext, useRef, createElement } from 'react';
import { MuiPickersAdapterContext } from './LocalizationProvider.js';
import { useTheme } from '@material-ui/core/styles';
import getThemeProps from '@material-ui/styles/getThemeProps';

// TODO uncomment when syntax will be allowed by next babel
function checkUtils(utils)
/* :asserts utils is MuiPickersAdapter */
{
  if (!utils) {
    throw new Error('Can not find utils in context. It looks like you forgot to wrap your component in LocalizationProvider, or pass dateAdapter prop directly.');
  }
}

function useUtils() {
  var utils = useContext(MuiPickersAdapterContext);
  checkUtils(utils);
  return utils;
}
function useNow() {
  var utils = useUtils();
  var now = useRef(utils.date());
  return now.current;
}

function withDefaultProps(_ref, Component) {
  var name = _ref.name;
  var componentName = name.replace('Mui', '');

  var WithDefaultProps = function WithDefaultProps(props) {
    var theme = useTheme();
    var propsWithDefault = getThemeProps({
      props: props,
      theme: theme,
      name: name
    });
    Component.displayName = componentName;
    return /*#__PURE__*/createElement(Component, propsWithDefault);
  };

  WithDefaultProps.displayName = "WithDefaultProps(".concat(componentName, ")");
  return WithDefaultProps;
}

var DIALOG_WIDTH = 320;
var DIALOG_WIDTH_WIDER = 325;
var VIEW_HEIGHT = 358;
var DAY_SIZE = 36;
var DAY_MARGIN = 2;
var IS_TOUCH_DEVICE_MEDIA = '@media (pointer: fine)';

export { DAY_MARGIN as D, IS_TOUCH_DEVICE_MEDIA as I, VIEW_HEIGHT as V, useNow as a, DAY_SIZE as b, DIALOG_WIDTH as c, DIALOG_WIDTH_WIDER as d, useUtils as u, withDefaultProps as w };
